go/vt/vttablet/grpctmclient/cached_client.go

```diff
@@ -0,0 +1,340 @@
+
+// Less is part of the sort.Interface interface and is used by container/heap
+// functions.
+func (queue cachedConns) Less(i, j int) bool {
+	left, right := queue[i], queue[j]
+	if left.refs == right.refs {
+		// break ties by access time.
+		// more stale connections have higher priority for removal
+		// this condition is equvalent to:
+		//		left.lastAccessTime <= right.lastAccessTime
+		return !left.lastAccessTime.After(right.lastAccessTime)
+	}
+
+	// connections with fewer refs have higher priority for removal
+	return left.refs < right.refs
+}
+
+// Swap is part of the sort.Interface interface and is used by container/heap
+// functions.
+func (queue cachedConns) Swap(i, j int) {
+	queue[i], queue[j] = queue[j], queue[i]
+	queue[i].index = i
+	queue[j].index = j
+}
+
+// Push is part of the container/heap.Interface interface.
+func (queue *cachedConns) Push(x interface{}) {
+	n := len(*queue)
+	conn := x.(*cachedConn)
+	conn.index = n
+	*queue = append(*queue, conn)
+}
+
+// Pop is part of the container/heap.Interface interface.
+func (queue *cachedConns) Pop() interface{} {
+	old := *queue
+	n := len(old)
+	conn := old[n-1]
+	old[n-1] = nil  // avoid memory leak
+	conn.index = -1 // for safety
+	*queue = old[0 : n-1]
+
+	return conn
+}
+
+type cachedConnDialer struct {
+	m            sync.RWMutex
+	conns        map[string]*cachedConn
+	qMu          sync.Mutex
+	queue        cachedConns
+	connWaitSema *sync2.Semaphore
+}
+
+var dialerStats = struct {
+	ConnReuse            *stats.Gauge
+	ConnNew              *stats.Gauge
+	DialTimeouts         *stats.Gauge
+	DialTimings          *stats.Timings
+	EvictionQueueTimings *stats.Timings
+}{
+	ConnReuse:    stats.NewGauge("tabletmanagerclient_cachedconn_reuse", "number of times a call to dial() was able to reuse an existing connection"),
+	ConnNew:      stats.NewGauge("tabletmanagerclient_cachedconn_new", "number of times a call to dial() resulted in a dialing a new grpc clientconn"),
+	DialTimeouts: stats.NewGauge("tabletmanagerclient_cachedconn_dial_timeouts", "number of context timeouts during dial()"),
+	DialTimings:  stats.NewTimings("tabletmanagerclient_cachedconn_dialtimings", "timings for various dial paths", "path", "rlock_fast", "sema_fast", "sema_poll"),
+	// TODO: add timings for heap operations (push, pop, fix)
+}
+
+// NewCachedConnClient returns a grpc Client using the priority queue cache
+// dialer implementation.
+func NewCachedConnClient(capacity int) *Client {
+	dialer := &cachedConnDialer{
+		conns:        make(map[string]*cachedConn, capacity),
+		queue:        make(cachedConns, 0, capacity),
+		connWaitSema: sync2.NewSemaphore(capacity, 0),
+	}
+
+	heap.Init(&dialer.queue)
+	return &Client{dialer}
+}
+
+var _ dialer = (*cachedConnDialer)(nil)
+
+func (dialer *cachedConnDialer) dial(ctx context.Context, tablet *topodatapb.Tablet) (tabletmanagerservicepb.TabletManagerClient, io.Closer, error) {
+	start := time.Now()
+
+	addr := getTabletAddr(tablet)
+	dialer.m.RLock()
```

Comment:
It might simplify the locking/unlocking to break the following few lines out into their own function, so you can just defer right after locking